include/nbl/asset/ICPUBuffer.h

```diff
@@ -107,7 +99,37 @@ class ICPUBuffer : public asset::IBuffer, public asset::IAsset
             return true;
         }
 
+        bool equals(const IAsset* _other) const override
+		{
+			auto* other = static_cast<const ICPUBuffer*>(_other);
+			return compatible(other) && (m_creationParams.usage.value == other->m_creationParams.usage.value);
+		}
+
+
+		bool canBeRestoredFrom(const IAsset* _other) const override
+		{
+			auto* other = static_cast<const ICPUBuffer*>(_other);
+			return compatible(other);
+		}
+
+		size_t hash(std::unordered_map<IAsset*, size_t>* temporary_hash_cache = nullptr) const override
+		{
+			size_t seed = AssetType;
+			core::hash_combine(seed, m_creationParams.size);
+			core::hash_combine(seed, m_creationParams.usage.value);
+			return seed;
+		}
+
     protected:
+
+	    bool compatible(const IAsset* _other) const override {
+		    if (IAsset::compatible(_other)) {
+			    auto* other = static_cast<const ICPUBuffer*>(_other);
+                return (m_creationParams.size == other->m_creationParams.size);
+		    }
+		    return false;

```

Comment:
if you followed the `_impl` pattern, youu would not need to remember to call `IAsset::compatible` and you'd be sure that th function is called with compat asset already

Reply:
https://discord.com/channels/593902898015109131/1078226704092381194/1083069244650045600